cores/esp8266/StackThunk.cpp

```diff
@@ -48,7 +50,14 @@ void stack_thunk_add_ref()
 {
   stack_thunk_refcnt++;
   if (stack_thunk_refcnt == 1) {
+    DBG_MMU_PRINTF("\nStackThunk malloc(%u)\n", _stackSize * sizeof(uint32_t));
+    // The stack must be in DRAM, or an Soft WDT will follow. Not sure why,
+    // maybe too much time is consumed with the non32-bit exception handler.
+    // Also, interrupt handling on an IRAM stack would be very slow.
+    // Strings on the stack would be very slow to access as well.
+    HeapSelectDram ephemeral;
```

Comment:
This makes me think about interrupt safety in the ephemeral constructor... are we 100% sure the heap change is interrupt safe, even from itself?

Reply:
I think it is safe as long as an ISR does not get interrupts enabled. 
There was a window with the non32-exception handler; however, I think I have that one closed.
ISRs should not try and change heaps. Even then all should be well if they put it back the way it was. On return, everything should be the same. The foreground thread will continue with the same information that was there before the interrupt.

Changing the heap involves changing a single variable. From the umm_mallac side, that variable is used to load a context pointer by index. From there it stays in registers or the current stack as the request is processed. With a single variable to reference for heap selection, I am thinking it is unlikely that umm_malloc can be called with things in an unusable transition state.

Additionally from the umm_malloc side, it can decide as in the ISR case, to ignore the current context and locally load the context for IRAM. No need to touch the current heap index variable.

Async logic without locks is always tricky. I keep asking the question: "Have I missed an angle?"

One concern I have about using IRAM instead of DRAM would be DMA. Do we have any DMA operation taking place with the ESP8266? I would expect the Network buffer transfers to be DMA. For TCP transfers I believe lwIP is configured to always copy sends to its buffers which would be in DRAM so that should not be a problem. I am less familiar with TCP receive and the UDP configuration options.

Reply:
Network (wifi) uses `malloc()` too. I believe it only happens at `yields()` and (end of) `loop()`.
One/I should conduct some tests under heavy network load and see what happens.
If a network allocation happens in IRAM without the unaligned handler, bad things may happen.

Reply:
Then, I think the changes that were made here should keep Network malloc()s in DRAM.
https://github.com/esp8266/Arduino/blob/a3bf35cff7abc0bd89090601b112efc218c07f33/cores/esp8266/core_esp8266_main.cpp#L207-L216

Any SYS callbacks that changes the Heap must restore before they return.